import { Categories } from "../models/Category.models";

export const create = async (req, res) => {
  console.log("ğŸš€ ~ create ~ req:", req);
  const category = new Categories(req.body);
  console.log("ğŸš€ ~ create ~ req:", req);
  category.manager = req.user.id;
  console.log("ğŸš€ ~ create ~ req:", req);
  try {
    await category.save();
    res.status(201).json({ message: "CategorÃ­a creada correctamente" });
  } catch (error) {
    console.log(error);
    res.status(500).json({ message: "Error al crear la categoria" });
  }
};

export const getAllCategories = async (req, res) => {
  console.log("ğŸš€ ~ getAllCategories ~ req:", req.user.id);

  try {
    const categories = await Categories.find({
      $or: [
        {
          manager: {
            $in: req.user.id,
          },
        },
      ],
    });
    res.json(categories);
  } catch (error) {
    console.log(error);
    res.status(500).json({ message: "Error al obtener las categorÃ­as" });
  }
};

export const getCategoryById = async (req, res) => {
  const id = req.params.id;

  try {
    const category = await Categories.findById(id).populate("services");
    if (!category) {
      return res.status(404).json({ message: "CategorÃ­a no encontrada" });
    }
    if (category.manager.toString() !== req.user.id) {
      return res.status(404).json({ message: "AcciÃ³n no valida" });
    }

    res.json(category);
  } catch (error) {
    console.log(error);
    res.status(500).json({ message: "Error al obtener la categorÃ­a" });
  }
};

export const updateCategory = async (req, res) => {
  console.log("ğŸš€ ~ updateCategory ~ req:", req);
  const id = req.params.id;
  console.log("ğŸš€ ~ updateCategory ~ req:", req);
  const data = req.body;
  console.log("ğŸš€ ~ updateCategory ~ req:", req);

  try {
    const category = await Categories.findById(id);
    if (!category) {
      return res.status(404).json({ message: "CategorÃ­a no encontrada" });
    }
    if (category.manager.toString() !== req.user.id) {
      console.log("ğŸš€ ~ updateCategory ~ req:", req);
      return res.status(404).json({ message: "AcciÃ³n no valida" });
    }
    category.name = data.name;
    await category.save();
    res.json({ message: "CategorÃ­a fue actualizada" });
  } catch (error) {
    console.log(error);
    res.status(500).json({ message: "Error al actualizar la categorÃ­a" });
  }
};

export const deleteCategory = async (req, res) => {
  console.log("ğŸš€ ~ deleteCategory ~ req:", req);
  const id = req.params.id;
  console.log("ğŸš€ ~ deleteCategory ~ req:", req);

  try {
    const category = await Categories.findById(id);
    if (!category) {
      return res.status(404).json({ message: "CategorÃ­a no encontrada" });
    }
    if (category.manager.toString() !== req.user.id) {
      console.log("ğŸš€ ~ deleteCategory ~ req:", req);
      return res.status(404).json({ message: "AcciÃ³n no valida" });
    }
    await category.deleteOne();
    res.json({ message: "CategorÃ­a fue eliminada" });
  } catch (error) {
    console.log(error);
    res.status(500).json({ message: "Error al eliminar la categorÃ­a" });
  }
};
